# Github

https://github.com/max-eisenberg/Project

# Project Description
Strava is a popular app used by athletes to track workouts. In our project we give users insight to their current performance and use activity data of users that are outperform the user to give reccomendations on how to improve. The segments feature of strava takes stretches from user activity to compare to other users who also complete that strech in one of their logged activities. Any segment in which a user performs amongst the top ten best is stored in the profile under top tens.



# General Approach

This project uses a webscraping approach which required the use of scrapy and selenium. While strava does have an API available to the public, it is severly limited and requires authentification by the user before taking any activity data. Given that our project relies so heavily on a analyzing competitor activities, the API was not of much use to us. This project looks at anyone who outperformed the user in their top tens and use their activity data to help the user improve. With out sraper we first logged into strava, directed to the top tens page, visited all of the top ten segments, got all of the athletes who beat the user in each top ten, then directed to their athelte page to collect all of their activity for the month leading up to their achievement.


# Our Results

Our webscraping gives us a lot of activity data, which we clean up before using it for analysis.

In [9]:
import pandas as pd
import numpy as np
df = pd.read_csv('individual_activities.csv')
df

Name  Distance Elevation     Time
0          Jae Kim  30.57 mi  2,257 ft   1h 56m
1          Jae Kim  48.77 mi  2,172 ft   3h 40m
2          Jae Kim  33.18 mi  1,306 ft   2h 11m
3          Jae Kim  31.78 mi  1,150 ft   2h 21m
4          Jae Kim  27.99 mi  1,093 ft   1h 58m
...            ...       ...       ...      ...
4229  Sam Boardman  37.55 mi  1,634 ft    2h 7m
4230  Sam Boardman  36.55 mi    272 ft   2h 13m
4231  Sam Boardman  13.81 mi    571 ft  48m 50s
4232  Sam Boardman  40.32 mi  1,056 ft   2h 32m
4233  Sam Boardman  29.10 mi    735 ft   1h 47m

[4234 rows x 4 columns]

Raw data

---

In [10]:
# Getting rid of duplicate activities and non-bike rides
df = df.drop_duplicates()
df = df[df['Time'].str.endswith('m', 's')]
df = df[df['Distance'].str.endswith('mi')]
df = df[df['Elevation'].str.endswith('ft')]

# Casting Data into integers
df['Distance'] = df['Distance'].apply(lambda x: x[:-3] if len(x) > 2 else x)
df['Distance'] = df['Distance'].astype(float)

df['Elevation'] = df['Elevation'].apply(lambda x: x[:-3] if len(x) > 2 else x)
df['Elevation'] = df['Elevation'].str.replace(',', '').astype(int)

# Transforming time into minutes, to perform arithmetic
hours = np.empty((df.shape[0],1))
hrs = []
minutes = np.empty((df.shape[0],1))
mins = []

for i in df['Time']:
    if 'h' in i:
        loc = i.find('h')
        hrs.append(int(i[:loc]))
        loc_min = i.find('m')
        mins.append(int(i[loc + 2:loc_min]))
    else:
        loc = i.find('m')
        mins.append(int(i[:loc]))

hours = np.array(hrs)
minutes = np.array(mins)
time_minutes = (hours * 60) + minutes

df = df.drop('Time', axis=1)
df['Time (min)'] = time_minutes

# Renaming the columns
df = df.rename(columns={'Distance': 'Distance (mi)', 'Elevation': 'Elevation (ft)', 'Name' : 'Athlete Name' })

df

Athlete Name  Distance (mi)  Elevation (ft)  Time (min)
0          Jae Kim          30.57            2257         116
1          Jae Kim          48.77            2172         220
2          Jae Kim          33.18            1306         131
3          Jae Kim          31.78            1150         141
4          Jae Kim          27.99            1093         118
...            ...            ...             ...         ...
4227  Sam Boardman          33.05            2175         103
4229  Sam Boardman          37.55            1634         127
4230  Sam Boardman          36.55             272         133
4232  Sam Boardman          40.32            1056         152
4233  Sam Boardman          29.10             735         107

[2213 rows x 4 columns]

*Clean data, statistic turned into integers, no bike-ride activities removed*

---

Then we find monthly totals for each athlete: 

In [11]:
# Calculating the total monthly distances

df['Distance (mi)'] = df.groupby('Athlete Name')['Distance (mi)'].transform('sum')
df['Elevation (ft)'] = df.groupby('Athlete Name')['Elevation (ft)'].transform('sum')
df['Time (min)'] = df.groupby('Athlete Name')['Time (min)'].transform('sum')
df = df.rename(columns={'Distance (mi)': 'Monthly Distance (mi)', 
                        'Elevation (ft)': 'Monthly Elevation (ft)', 
                        'Time (min)': 'Monthly Time (min)'})

df = df.drop_duplicates()

# Putting time into hours
df['Monthly Time (min)'] = df['Monthly Time (min)'].apply(lambda x: x/60)
df = df.rename(columns={'Monthly Time (min)': 'Monthly Time (hours)'})

df



Athlete Name  Monthly Distance (mi)  Monthly Elevation (ft)  \
0            Jae Kim                 397.37                   19768   
15      Jeff Johnson                2198.54                  188130   
34      Brook Sutt🍪n                 360.16                   33217   
51      Kerry Werner                 203.87                   10637   
60    Spencer Paxson                 408.94                   48279   
...              ...                    ...                     ...   
3994   Carl Anderson                 305.68                   41690   
4031    Nathan Lloyd                 474.34                   48040   
4043  Tyler Schwartz                1588.53                  107353   
4072      Safa Brian                 733.28                   75964   
4093    Sam Boardman                9588.19                  445221   

      Monthly Time (hours)  
0                28.300000  
15              141.116667  
34               28.566667  
51                9.766667  
60               42.650000  
...                    ...  
3994             61.600000  
4031             32.466667  
4043             90.550000  
4072             46.666667  
4093            404.800000  

[101 rows x 4 columns]

*Average stats found for each athelte, duplicates removed, time put back into hours*

---

In [12]:
cleaned_data = df
average_distance = cleaned_data['Monthly Distance (mi)'].mean().round(2)
average_elevation = cleaned_data['Monthly Elevation (ft)'].mean().round(2)
average_time = cleaned_data['Monthly Time (hours)'].mean().round(2)

average_weekly_distance = (average_distance/4).round(2)
average_weekly_time = (average_time/4).round(2)

print(f'On Average, people that did better than you on your top 10 segments rode {average_distance} miles \na month, climbed {average_elevation} feet per month, and spent {average_time} hours training. This type \nof exertion amounts to roughly {average_weekly_distance} miles and {average_weekly_time} hours of training per week. \nGo get em champ!')


On Average, people that did better than you on your top 10 segments rode 937.25 miles 
a month, climbed 69084.92 feet per month, and spent 58.04 hours training. This type 
of exertion amounts to roughly 234.31 miles and 14.51 hours of training per week. 
Go get em champ!


### Arch nemesis
 For this section we found the most reoccuring athletes to show who your "Arch Nemesis" is:

In [13]:
links = pd.read_csv("Interval_pages.csv")
links['Athlete ID'] = links['Athlete page'].str.extract(r'/athletes/(\d+)')
links = links.dropna()
from collections import Counter
link_counts = Counter(links["Athlete ID"])
occurrences_df = pd.DataFrame(list(link_counts.items()), columns=["Athlete ID", "Occurrences"])
occurrences_df = occurrences_df.sort_values(by="Occurrences", ascending=False)
occurrences_df = occurrences_df.reset_index(drop=True)

print("After looking through your top ten achievements we found all user who beat you in your \ntop ten segments. Consider there people your \"Arch Nemisis.\" We recommend \nchecking out their profile to see how you can improve:  ")

print(" ")
print(" ")

print(f'Your greatest arch nemisis beat you {occurrences_df.loc[0, "Occurrences"]} times in your current top ten segments. \nFind their profile here: {"https://www.strava.com/athletes/"+occurrences_df.loc[0,"Athlete ID"]}')

print(" ")

print(f'Your second greatest arch nemisis beat you {occurrences_df.loc[1, "Occurrences"]} times in your current top ten segments.\nFind their profile here: {"https://www.strava.com/athletes/"+occurrences_df.loc[1,"Athlete ID"]}')

print(" ")

print(f'Your third greatest arch nemisis beat you {occurrences_df.loc[2, "Occurrences"]} times in your current top ten segments. \nFind their profile here: {"https://www.strava.com/athletes/"+ occurrences_df.loc[2,"Athlete ID"]}')

After looking through your top ten achievements we found all user who beat you in your 
top ten segments. Consider there people your "Arch Nemisis." We recommend 
checking out their profile to see how you can improve:  
 
 
Your greatest arch nemisis beat you 7 times in your current top ten segments. 
Find their profile here: https://www.strava.com/athletes/14197582
 
Your second greatest arch nemisis beat you 7 times in your current top ten segments.
Find their profile here: https://www.strava.com/athletes/5875016
 
Your third greatest arch nemisis beat you 7 times in your current top ten segments. 
Find their profile here: https://www.strava.com/athletes/497379




<figure>
    <center><a href="https://www.strava.com/athletes/14197582"><img src="joe_strava.jpeg"
         alt="Joe" class="center" width="150" height="auto"><center>
    <figcaption> <a href="https://www.strava.com/athletes/14197582" target="_blank"><center>Joe Emmerling<center></a>
</figure>

<figure>
    <center><a href="https://www.strava.com/athletes/5875016"><img src="danny_strava.png"
         alt="Tony" class="center" width="150" height="auto"><center>
    <figcaption> <a href="https://www.strava.com/athletes/5875016" target="_blank"><center>Danny Finneran<center></a>
</figure>

<figure>
    <center><a href="https://www.strava.com/athletes/497379"><img src="tony_strava2.jpeg"
         alt="Tony" class="center" width="150" height="auto"><center>
    <figcaption> <a href="https://www.strava.com/athletes/497379" target="_blank"><center>Tony Manzella<center></a>
</figure>


*Arch nemesis analysis*

---

### Max's Data
With the API we were able to get Max's Data which is helpful for comparison to his competitors

2022 Weekly Mileage

In [14]:
df = pd.read_csv('max_data.csv')
df['Year']= df["Date"].str[:4]
df['Month']= df["Date"].str[5:7]

import datetime
week = [] 
for i in df['Date']:
    month = int(i[:4])
    day = int(i[5:7])
    year = int(i[8:])
    week.append(datetime.date(month,day,year).strftime("%V"))
df['Week']=week

#Weekly mileage
df_2022 = df[df['Year'] == '2022'].sort_values('Week')
import plotly.express as px
import plotly.io as pio
pio.renderers.default="iframe"
# Here we use a column with categorical data
fig = px.histogram(df_2022, x="Week", y= "Distance (mi)")
fig.show()

2022 Monthly Mileage

In [15]:
#Monthly mileage
import plotly.express as px
# Here we use a column with categorical data
fig = px.histogram(df_2022, x="Month", y= "Distance (mi)")
fig.show()

# The Proccess

## Flow of work

With our webscrpaing approach we followed this general flow:
- Log in to strava on scrapy
- Direct to strava athlete top ten pages 
- Get links to all top ten segments
- Direct to athlete page for all athletes that beat user in that segment
    - Specifically we direct to the athelte page which displays all of the activities from the month before that athlete placed in the the top ten segment, ahead of the user.
- Log in through selenium
- Gather all activiy data from the links for given time period of each athlete, store. 

## The Scraper

- First we compiled our scraper to go into the strava website and through the top ten segments gather all of the athletes ahead of the user and find their activity page corresponding to the month before their acheivement.
- Then we logged into strava with selenium and looped through all of the athlete activity pages and stored their activity data for the month.
- We also stored the name on the activity to later be able to throw out other user's group activities that are also present on the page.




In [16]:
import scrapy
import selenium
from selenium import webdriver
from scrapy.http import FormRequest
from scrapy_selenium import SeleniumRequest
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
import time

class StravaScraper(scrapy.Spider):
    name = "test"
    athlete = '16735685'
    segments =[]
    
    allowed_domains = ['strava.com']
    start_urls = ['https://www.strava.com/login']
    

    def __init__(self):
        self.date_dict = {'Jan':'01','Feb':'02','Mar':'03','Apr':'04','May':'05','Jun':'06','Jul':'07','Aug':'08',
                          'Sep':'09','Oct':'10','Nov':'11','Dec':'12'}
        self.month_before = {'Jan':'12','Feb':'01','Mar':'02','Apr':'03','May':'04','Jun':'05','Jul':'06','Aug':'07',
                          'Sep':'08','Oct':'09','Nov':'10','Dec':'11'}


    def parse(self, response):
        token = response.xpath('//*[@name="csrf-token"]/@content').get()
        return FormRequest.from_response(response,
                                        formdata={
                                            'authenticity_token': token,
                                            
                                            'email': 'sashaprs@gmail.com',
                                            'password': 'PIC16BProject',
                                        },
                                        #dont_filter=True,
                                        #eta={'dont_redirect': True, 'handle_httpstatus_list': [302]},
                                        callback=self.parse_after_login)
    

    def parse_after_login(self, response):
        #Try except statement for if user has no top tens
        top_ten_page = f'https://www.strava.com/athletes/{self.athlete}/segments/leader?top_tens=true'
    
        yield scrapy.Request(url=top_ten_page, callback=self.parse_top_tens)

    def parse_top_tens(self, response):
        top_tens = response.css('table.my-segments tbody tr td a::attr(href)').getall()
        self.segments.extend(top_tens)
        next_page = response.xpath('//li[@class="next_page"]/a[@rel="next"]/@href').get()
        if(next_page):
            next_page_url = 'https://www.strava.com' + next_page
        
            yield scrapy.Request(url=next_page_url, callback = self.parse_top_tens)
            
        for top_ten in self.segments:

            if '/segments/' in top_ten:
            # yield {"top ten": 'https://www.strava.com' + top_ten}
                top_ten = 'https://www.strava.com' + top_ten
                yield scrapy.Request(url=top_ten, callback = self.parse_leaderboard)
    
    def parse_leaderboard(self, response):
        self.counter +=1
        self.url_list = [] 
        athlete_pages = response.css('td.athlete.track-click a::attr(href)').getall()
        dates = response.css('a[href^="/segment_efforts/"]::text').getall()
        #This get month of acheivement, we want month before acheivement
        edited_dates = []

        #New code: Defined new dictionary that gives month before, if month before is december, also subtracts one from year.
        for d in dates:
            month = self.month_before[d[:3]]
            if month == "12":
                year = int(d[-4:])
                year = str(year-1)
            else:
                year = d[-4:]
            edited_dates.append(year+month)

        athlete = '16735685'  # Ensure this is a string

        athletes_data = {}  # Dictionary to store athlete URLs and corresponding dates

        for i, athlete_url in enumerate(athlete_pages):
            if athlete_url.endswith(f'/athletes/{athlete}'):
                break
            athlete_url = 'https://www.strava.com' + athlete_url
            athletes_data[athlete_url] = edited_dates[i]

        for athlete_url, date in athletes_data.items():
            interval = str(date)
        
            year_offset = str(2023 - int(year))
            athlete_url = f'{athlete_url}#interval?interval={interval}&interval_type=month&chart_type=miles&year_offset={year_offset}'
            yield {'Athlete': athlete_url}
                
            


### Selenium

In [ ]:
import scrapy
import selenium
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
import time
import pandas as pd

activities = pd.DataFrame()
test = ['https://www.strava.com/athletes/29127886#interval_type?chart_type=miles&interval_type=month&interval=202012&year_offset=3','https://www.strava.com/athletes/5838343#interval_type?chart_type=miles&interval_type=month&interval=201803&year_offset=5','https://www.strava.com/athletes/28626466#interval_type?chart_type=miles&interval_type=month&interval=202205&year_offset=1']

urls = pd.read_csv('Interval_pages.csv')
urls_list = urls['Athlete page']
executable_path="/Users/sasha/Documents/PIC16B/strava_scraper/strava_scraper/spiders/chromedriver"
driver = webdriver.Chrome(executable_path=executable_path)
        

driver.get('https://www.strava.com/login')
email_field = driver.find_element_by_id('email')
email_field.send_keys('sashaprs@gmail.com')
password_field = driver.find_element_by_id('password')
password_field.send_keys('PIC16BProj')
login_button = driver.find_element_by_id('login-button')
login_button.click()

for url in urls_list:
    temp=[]
    activity_stats = [] 
    names = []
    driver.get(url)
    driver.implicitly_wait(10)
    try:
        athlete = driver.find_element(By.CSS_SELECTOR, "h1.text-title1.athlete-name").text
        name = driver.find_elements(By.XPATH("//a[@data-testid='owners-name'"))
        for i in name:
            names.append(i.text)
        activity_values = driver.find_elements(By.CLASS_NAME, "------packages-ui-Stat-Stat-module__statValue--phtGK")
        for i in activity_values:
            temp.append(i.text)
        distance = temp[::3]
        elevation = temp[1::3]
        time = temp[2::3]
        df = pd.DataFrame({'Athlete':athlete,'Name':names,'Distance': distance, 'Elevation': elevation, 'Time': time})
        activities = pd.concat([activities, df], axis=0)
    except:
        pass
activities.to_csv('individual_acitivities.csv', index=False)


## Data Processing and Analysis
### Training Reccomendations

This gave us our activity data. The raw data was 4234 activities.

In [18]:
import pandas as pd
import numpy as np
df = pd.read_csv('individual_activities.csv')
df

Name  Distance Elevation     Time
0          Jae Kim  30.57 mi  2,257 ft   1h 56m
1          Jae Kim  48.77 mi  2,172 ft   3h 40m
2          Jae Kim  33.18 mi  1,306 ft   2h 11m
3          Jae Kim  31.78 mi  1,150 ft   2h 21m
4          Jae Kim  27.99 mi  1,093 ft   1h 58m
...            ...       ...       ...      ...
4229  Sam Boardman  37.55 mi  1,634 ft    2h 7m
4230  Sam Boardman  36.55 mi    272 ft   2h 13m
4231  Sam Boardman  13.81 mi    571 ft  48m 50s
4232  Sam Boardman  40.32 mi  1,056 ft   2h 32m
4233  Sam Boardman  29.10 mi    735 ft   1h 47m

[4234 rows x 4 columns]

To clean up the data we changed the datatype of the statistics to integers, including changing the time to minutes to help with arithmetic later. We also removed any activites that were not bike rides. This resulted in 2213 rows of data.

In [19]:
# Getting rid of duplicate activities and non-bike rides
df = df.drop_duplicates()
df = df[df['Time'].str.endswith('m', 's')]
df = df[df['Distance'].str.endswith('mi')]
df = df[df['Elevation'].str.endswith('ft')]

# Casting Data into integers
df['Distance'] = df['Distance'].apply(lambda x: x[:-3] if len(x) > 2 else x)
df['Distance'] = df['Distance'].astype(float)

df['Elevation'] = df['Elevation'].apply(lambda x: x[:-3] if len(x) > 2 else x)
df['Elevation'] = df['Elevation'].str.replace(',', '').astype(int)

# Transforming time into minutes, to perform arithmetic
hours = np.empty((df.shape[0],1))
hrs = []
minutes = np.empty((df.shape[0],1))
mins = []

for i in df['Time']:
    if 'h' in i:
        loc = i.find('h')
        hrs.append(int(i[:loc]))
        loc_min = i.find('m')
        mins.append(int(i[loc + 2:loc_min]))
    else:
        loc = i.find('m')
        mins.append(int(i[:loc]))

hours = np.array(hrs)
minutes = np.array(mins)
time_minutes = (hours * 60) + minutes

df = df.drop('Time', axis=1)
df['Time (min)'] = time_minutes

# Renaming the columns
df = df.rename(columns={'Distance': 'Distance (mi)', 'Elevation': 'Elevation (ft)', 'Name' : 'Athlete Name' })

df

Athlete Name  Distance (mi)  Elevation (ft)  Time (min)
0          Jae Kim          30.57            2257         116
1          Jae Kim          48.77            2172         220
2          Jae Kim          33.18            1306         131
3          Jae Kim          31.78            1150         141
4          Jae Kim          27.99            1093         118
...            ...            ...             ...         ...
4227  Sam Boardman          33.05            2175         103
4229  Sam Boardman          37.55            1634         127
4230  Sam Boardman          36.55             272         133
4232  Sam Boardman          40.32            1056         152
4233  Sam Boardman          29.10             735         107

[2213 rows x 4 columns]

We then condensed the dataframe to find montly totals for each rider. Now there is one column for each athlete with their monthly bike-ride totals.

In [20]:
# Calculating the total monthly distances

df['Distance (mi)'] = df.groupby('Athlete Name')['Distance (mi)'].transform('sum')
df['Elevation (ft)'] = df.groupby('Athlete Name')['Elevation (ft)'].transform('sum')
df['Time (min)'] = df.groupby('Athlete Name')['Time (min)'].transform('sum')
df = df.rename(columns={'Distance (mi)': 'Monthly Distance (mi)', 
                        'Elevation (ft)': 'Monthly Elevation (ft)', 
                        'Time (min)': 'Monthly Time (min)'})

df = df.drop_duplicates()

# Putting time into hours
df['Monthly Time (min)'] = df['Monthly Time (min)'].apply(lambda x: x/60)
df = df.rename(columns={'Monthly Time (min)': 'Monthly Time (hours)'})

df

Athlete Name  Monthly Distance (mi)  Monthly Elevation (ft)  \
0            Jae Kim                 397.37                   19768   
15      Jeff Johnson                2198.54                  188130   
34      Brook Sutt🍪n                 360.16                   33217   
51      Kerry Werner                 203.87                   10637   
60    Spencer Paxson                 408.94                   48279   
...              ...                    ...                     ...   
3994   Carl Anderson                 305.68                   41690   
4031    Nathan Lloyd                 474.34                   48040   
4043  Tyler Schwartz                1588.53                  107353   
4072      Safa Brian                 733.28                   75964   
4093    Sam Boardman                9588.19                  445221   

      Monthly Time (hours)  
0                28.300000  
15              141.116667  
34               28.566667  
51                9.766667  
60               42.650000  
...                    ...  
3994             61.600000  
4031             32.466667  
4043             90.550000  
4072             46.666667  
4093            404.800000  

[101 rows x 4 columns]

Finally, we found the average of all the athletes statistic which gave us our training reccomendation for both the month and per week:

In [21]:
cleaned_data = df
average_distance = cleaned_data['Monthly Distance (mi)'].mean().round(2)
average_elevation = cleaned_data['Monthly Elevation (ft)'].mean().round(2)
average_time = cleaned_data['Monthly Time (hours)'].mean().round(2)

average_weekly_distance = (average_distance/4).round(2)
average_weekly_time = (average_time/4).round(2)

print(f'On Average, people that did better than you on your top 10 segments rode {average_distance} miles a month, climbed {average_elevation} feet per month, \nand spent {average_time} hours training. This type of exertion amounts to roughly {average_weekly_distance} miles and {average_weekly_time} hours of training per week. \nGo get em champ!')


On Average, people that did better than you on your top 10 segments rode 937.25 miles a month, climbed 69084.92 feet per month, 
and spent 58.04 hours training. This type of exertion amounts to roughly 234.31 miles and 14.51 hours of training per week. 
Go get em champ!


### Arch nemesis

For the arch nemesis section we used the list of links we used for selinium to find the users that beat the user most often. Since the link is specific to the timeframe we isolated the athlete ID before find the top three most occuring users:

In [22]:
links = pd.read_csv("Interval_pages.csv")
links['Athlete ID'] = links['Athlete page'].str.extract(r'/athletes/(\d+)')
links = links.dropna()
from collections import Counter
link_counts = Counter(links["Athlete ID"])
occurrences_df = pd.DataFrame(list(link_counts.items()), columns=["Athlete ID", "Occurrences"])
occurrences_df = occurrences_df.sort_values(by="Occurrences", ascending=False)
occurrences_df = occurrences_df.reset_index(drop=True)

occurrences_df

Athlete ID  Occurrences
0     14197582            7
1      5875016            7
2       497379            7
3      5350307            6
4       571689            5
..         ...          ...
180   34799460            1
181   45025631            1
182     497009            1
183   65336496            1
184    2644812            1

[185 rows x 2 columns]

In [23]:

print("After looking through your top ten achievements we found all user who beat you in your \ntop ten segments. Consider there people your \"Arch Nemisis.\" We recommend \nchecking out their profile to see how you can improve:  ")

print(" ")
print(" ")

print(f'Your greatest arch nemisis beat you {occurrences_df.loc[0, "Occurrences"]} times in your current top ten segments. \nFind their profile here: {"https://www.strava.com/athletes/"+occurrences_df.loc[0,"Athlete ID"]}')

print(" ")

print(f'Your second greatest arch nemisis beat you {occurrences_df.loc[1, "Occurrences"]} times in your current top ten segments.\nFind their profile here: {"https://www.strava.com/athletes/"+occurrences_df.loc[1,"Athlete ID"]}')

print(" ")

print(f'Your third greatest arch nemisis beat you {occurrences_df.loc[2, "Occurrences"]} times in your current top ten segments. \nFind their profile here: {"https://www.strava.com/athletes/"+ occurrences_df.loc[2,"Athlete ID"]}')

After looking through your top ten achievements we found all user who beat you in your 
top ten segments. Consider there people your "Arch Nemisis." We recommend 
checking out their profile to see how you can improve:  
 
 
Your greatest arch nemisis beat you 7 times in your current top ten segments. 
Find their profile here: https://www.strava.com/athletes/14197582
 
Your second greatest arch nemisis beat you 7 times in your current top ten segments.
Find their profile here: https://www.strava.com/athletes/5875016
 
Your third greatest arch nemisis beat you 7 times in your current top ten segments. 
Find their profile here: https://www.strava.com/athletes/497379
